launcher/launcher.go

```diff
@@ -153,32 +174,44 @@ func (l *launcher) stopBeater() {
 
 	// By waiting to the wait group, it make sure that the old beater has really stopped
 	l.wg.Wait()
+	l.log.Info("Launcher shutted the Beater down gracefully")
 }
 
-func (l *launcher) waitForUpdates() error {
+// waitForFinish is the function that keeps Launcher up and running
+// It will finish for one of two reasons:
+// 1. The context is done, that probably means that the Stop function got called
+// 2. The beater run has failed and returned an error
+// Note that if the beater returned nil the launcher will keep running (and a new beater should be up again)
+func (l *launcher) waitForFinish() error {
 	for {
 		select {
 		case <-l.ctx.Done():
-			l.stopBeater()
 			return nil
 
 		case err := <-l.beaterErr:
 			if err != nil {
+				l.reloader.Stop()
 				return fmt.Errorf("beater returned an error:  %w", err)
 			}
+		}
+	}
+}
 
-		case update, ok := <-l.reloader.Channel():
-			if !ok {
-				return errors.New("Reloader channel closed")
-			}
-
-			go func() {
-				if err := l.configUpdate(update); err != nil {
-					l.beaterErr <- fmt.Errorf("failed to update Beater config: %w", err)
-					return
-				}
-			}()
+func (l *launcher) waitForUpdates() {
+	for {
+		update, ok := <-l.reloader.Channel()
+		if !ok {
+			l.log.Info("Reloader channel closed")
+			return
 		}
+
+		l.cfgmtx.Lock()
+		go func() {
+			defer l.cfgmtx.Unlock()
+			if err := l.configUpdate(update); err != nil {
+				l.beaterErr <- fmt.Errorf("failed to update Beater config: %w", err)
+			}
+		}()
```

Comment:
not sure why that requires another routine?
In the case of multiple configurations following one another I think Ir'll be fine to handle them one by one/

Reply:
By not opening a new routine I think we can even get rid of `cfgmtx`.
wdyt?

Reply:
I think you are right, I will give it a try! 😄